File: backend/android/app/src/main/java/com/fitness/trail/data/api/HealthAPIClient.kt
package com.fitness.trail.data.api

import android.content.Context
import com.fitness.trail.data.health.HealthSnapshot
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class HealthAPIClient private constructor(context: Context) {

    companion object {
        @Volatile
        private var instance: HealthAPIClient? = null

        fun shared(context: Context): HealthAPIClient {
            return instance ?: synchronized(this) {
                instance ?: HealthAPIClient(context.applicationContext).also { instance = it }
            }
        }
    }

    private val baseUrl = "http://localhost:8000"
    private val gson = Gson()
    private val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)

    private var token: String?
        get() = prefs.getString("auth_token", null)
        set(value) = prefs.edit().putString("auth_token", value).apply()

    val isLoggedIn: Boolean get() = token != null

    fun saveToken(token: String) {
        this.token = token
    }

    suspend fun register(email: String, password: String, displayName: String): String {
        val body = mapOf(
            "email" to email,
            "password" to password,
            "display_name" to displayName
        )
        val response: TokenResponse = post("/auth/register", body, authenticated = false)
        saveToken(response.accessToken)
        return response.accessToken
    }

    suspend fun login(email: String, password: String): String {
        val body = mapOf("email" to email, "password" to password)
        val response: TokenResponse = post("/auth/login", body, authenticated = false)
        saveToken(response.accessToken)
        return response.accessToken
    }

    suspend fun syncHealthData(snapshot: HealthSnapshot): SyncResponse {
        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        val body = mapOf(
            "log_date" to dateFormat.format(Date(snapshot.date)),
            "steps" to snapshot.steps,
            "hydration_ml" to snapshot.hydrationMl,
            "source" to "healthconnect"
        )
        return post("/sync", body, authenticated = true)
    }

    suspend fun fetchDashboard(): DashboardResponse = get("/dashboard")

    suspend fun fetchAchievements(): List<Achievement> = get("/achievements")

    private suspend fun <T> post(path: String, body: Map<String, Any>, authenticated: Boolean): T {
        val url = URL("$baseUrl$path")
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        if (authenticated && token != null) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }
        connection.doOutput = true

        connection.outputStream.use { it.write(gson.toJson(body).toByteArray()) }

        val responseCode = connection.responseCode
        val responseBody = connection.inputStream.bufferedReader().readText()

        if (responseCode !in 200..299) {
            val error = try {
                gson.fromJson(responseBody, APIError::class.java)?.detail
            } catch (e: Exception) { null }
            throw APIClientException(responseCode, error ?: "Unknown error")
        }

        return gson.fromJson(responseBody, (T::class.java))
    }

    private suspend fun <T> get(path: String): T {
        val url = URL("$baseUrl$path")
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        if (token != null) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }

        val responseCode = connection.responseCode
        val responseBody = connection.inputStream.bufferedReader().readText()

        if (responseCode !in 200..299) {
            val error = try {
                gson.fromJson(responseBody, APIError::class.java)?.detail
            } catch (e: Exception) { null }
            throw APIClientException(responseCode, error ?: "Unknown error")
        }

        return gson.fromJson(responseBody, (T::class.java))
    }
}

data class TokenResponse(val accessToken: String)

data class SyncResponse(
    val logDate: String,
    val steps: Int,
    val hydrationMl: Double,
    val stepsGoalMet: Boolean,
    val hydrationGoalMet: Boolean,
    val newAchievements: List<String>
)

data class DashboardResponse(
    val today: String,
    val steps: Int,
    val hydrationMl: Double,
    val stepGoal: Int,
    val hydrationGoalMl: Int,
    val stepsPct: Double,
    val hydrationPct: Double,
    val totalAchievements: Int
)

data class Achievement(
    val badgeId: String,
    val name: String,
    val desc: String,
    val icon: String,
    val earnedAt: String
) {
    val id: String get() = badgeId
}

data class APIError(val detail: String)

class APIClientException(val code: Int, override val message: String) : Exception(message)

File: backend/ios/HealthAPIClient.swift
import Foundation

class HealthAPIClient {
    
    static let shared = HealthAPIClient()
    
    private let baseURL = "http://172.25.147.19:8000"  // Your Mac's local IP
    private var token: String? {
        UserDefaults.standard.string(forKey: "auth_token")
    }
    
    // MARK: - Auth
    
    func register(email: String, password: String, displayName: String) async throws -> String {
        let body: [String: Any] = [
            "email": email,
            "password": password,
            "display_name": displayName
        ]
        let response: TokenResponse = try await post(path: "/auth/register", body: body, authenticated: false)
        saveToken(response.accessToken)
        return response.accessToken
    }
    
    func login(email: String, password: String) async throws -> String {
        let body: [String: Any] = ["email": email, "password": password]
        let response: TokenResponse = try await post(path: "/auth/login", body: body, authenticated: false)
        saveToken(response.accessToken)
        return response.accessToken
    }
    
    func saveToken(_ token: String) {
        UserDefaults.standard.set(token, forKey: "auth_token")
    }
    
    var isLoggedIn: Bool { token != nil }
    
    // MARK: - Health Sync
    
    @discardableResult
    func syncHealthData(snapshot: HealthSnapshot) async throws -> SyncResponse {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        
        let body: [String: Any] = [
            "log_date":      formatter.string(from: snapshot.date),
            "steps":         snapshot.steps,
            "hydration_ml":  snapshot.hydrationMl,
            "source":        "healthkit"
        ]
        return try await post(path: "/sync", body: body, authenticated: true)
    }
    
    // MARK: - Manual Hydration Sync

    /// Syncs a manual hydration entry. Fetches current HealthKit totals (which now
    /// include the just-saved manual entry) and posts to /sync with source "manual".
    @discardableResult
    func syncManualHydration(snapshot: HealthSnapshot) async throws -> SyncResponse {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"

        let body: [String: Any] = [
            "log_date":      formatter.string(from: snapshot.date),
            "steps":         snapshot.steps,
            "hydration_ml":  snapshot.hydrationMl,
            "source":        "manual"
        ]
        return try await post(path: "/sync", body: body, authenticated: true)
    }

    // MARK: - Dashboard
    
    func fetchDashboard() async throws -> DashboardResponse {
        return try await get(path: "/dashboard")
    }
    
    // MARK: - Achievements
    
    func fetchAchievements() async throws -> [Achievement] {
        return try await get(path: "/achievements")
    }
    
    // MARK: - Generic HTTP
    
    private func post<T: Decodable>(path: String, body: [String: Any], authenticated: Bool) async throws -> T {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if authenticated, let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)
        return try JSONDecoder.snakeCase.decode(T.self, from: data)
    }
    
    private func get<T: Decodable>(path: String) async throws -> T {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)
        return try JSONDecoder.snakeCase.decode(T.self, from: data)
    }
    
    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if !(200...299).contains(http.statusCode) {
            let message = (try? JSONDecoder().decode(APIError.self, from: data))?.detail ?? "Unknown error"
            throw APIClientError.serverError(http.statusCode, message)
        }
    }
}

// MARK: - Response Models

struct TokenResponse: Decodable {
    let accessToken: String
}

struct SyncResponse: Decodable {
    let logDate: String
    let steps: Int
    let hydrationMl: Double
    let stepsGoalMet: Bool
    let hydrationGoalMet: Bool
    let newAchievements: [String]
}

struct DashboardResponse: Decodable {
    let today: String
    let steps: Int
    let hydrationMl: Double
    let stepGoal: Int
    let hydrationGoalMl: Int
    let stepsPct: Double
    let hydrationPct: Double
    let totalAchievements: Int
}

struct Achievement: Decodable, Identifiable {
    var id: String { badgeId }
    let badgeId: String
    let name: String
    let desc: String
    let icon: String
    let earnedAt: String
}

struct APIError: Decodable {
    let detail: String
}

enum APIClientError: LocalizedError {
    case serverError(Int, String)
    var errorDescription: String? {
        switch self {
        case .serverError(let code, let msg): return "Error \(code): \(msg)"
        }
    }
}

// MARK: - Helpers

extension JSONDecoder {
    static var snakeCase: JSONDecoder {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }
}
